mne/preprocessing/ica.py

```diff
@@ -2079,3 +2079,239 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
+
+    if plot is True:
+        ttl = 'Template IC (Subj. ' + str(template[1]) + ')'
+        icas[template[0]].plot_components(picks=template[1],
+                                          ch_type=ch_type,
+                                          title=ttl)
+
+    # first run: use user-selected map
+    if isinstance(threshold, (int, float)):
+        try:
+            nt, mt, s, mx = find_max_corrs(all_maps, target, threshold)
+        except ValueError:
+            if threshold > 1:
+                logger.info("No component detected using find_outliers."
+                            " Consider using threshold='auto'")
+            return
+    elif len(threshold) > 1:
+        paths = [find_max_corrs(all_maps, target, t) for t in threshold]
+        # find iteration with highest avg correlation with target
+        nt, mt, s, mx = paths[np.argmax([path[2] for path in paths])]
+
+    # second run: use output from first run
+    if isinstance(threshold, (int, float)):
+        try:
+            nt, mt, s, mx = find_max_corrs(all_maps, nt, threshold)
+        except ValueError:
+            if threshold > 1:
+                logger.info("No component detected using find_outliers. "
+                            "Consider using threshold='auto'")
+            return
+    elif len(threshold) > 1:
+        paths = [find_max_corrs(all_maps, nt, t) for t in threshold]
+        # find iteration with highest avg correlation with target
+        nt, mt, s, mx = paths[np.argmax([path[1] for path in paths])]
+
+    nones = list(range(len(icas)))
+    new_icas, allmaps, indices, subjs = [], [], [], []
+    logger.info("Median correlation with constructed map: " + str(mt))
+    if plot:
+        logger.info("Displaying selected ICs per subject.")
+
+    for i, (ica, max_corr) in enumerate(zip(icas, mx)):
+        if inplace is False:
+            ica = deepcopy(ica)
+        if not hasattr(ica, 'labels'):
+            ica.labels = defaultdict(lambda: [])
+        try:
+            if isinstance(max_corr[0], np.ndarray):
+                max_corr = max_corr[0]
+            ica.labels[name] = list(set(list(max_corr) +
+                                    ica.labels[name]))
+            if plot:
+                allmaps.extend(get_ica_map(ica, components=max_corr))
+                subjs.extend([i] * len(max_corr))
+                indices.extend(max_corr)
+            nones.remove(i)
+        except IndexError:
+            pass
+        if inplace is False:
+            new_icas.append(ica)
+
+    if plot is True:
+        _plot_corrmap(allmaps, subjs, indices, ch_type, ica, name)
+
+    if nones:
+        logger.info("No maps selected for subject(s) " +
+                    ", ".join([str(x) for x in nones]) +
+                    ", consider a more liberal threshold.")
+    else:
+        logger.info("At least 1 IC detected for each subject.")
+
+    if inplace is True:
+        return
+    else:
+        return new_icas
```

Comment:
I'm not sure I fully understand the control flow, apparently you're using some recursion to get things done. Ideally the function should ultimately return new_icas or crash.